reactor-core/src/main/java/reactor/util/retry/RetryBackoffSpec.java

```diff
@@ -0,0 +1,566 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.util.retry;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+
+/**
+ * A builder for a retry strategy based on exponential backoffs, with fine grained options.
+ * Retry delays are randomized with a user-provided {@link #jitter(double)} factor between {@code 0.d} (no jitter)
+ * and {@code 1.0} (default is {@code 0.5}).
+ * Even with the jitter, the effective backoff delay cannot be less than {@link #minBackoff(Duration)}
+ * nor more than {@link #maxBackoff(Duration)}. The delays and subsequent attempts are executed on the
+ * provided backoff {@link #scheduler(Scheduler)}.
+ * <p>
+ * Only errors that match the {@link #filter(Predicate)} are retried (by default all),
+ * and the number of attempts can also limited with {@link #maxAttempts(long)}.
+ * When the maximum attempt of retries is reached, a runtime exception is propagated downstream which
+ * can be pinpointed with {@link reactor.core.Exceptions#isRetryExhausted(Throwable)}. The cause of
+ * the last attempt's failure is attached as said {@link reactor.core.Exceptions#retryExhausted(String, Throwable) retryExhausted}
+ * exception's cause. This can be customized with {@link #onRetryExhaustedThrow(BiFunction)}.
+ * <p>
+ * Additionally, to help dealing with bursts of transient errors in a long-lived Flux as if each burst
+ * had its own backoff, one can choose to set {@link #transientErrors(boolean)} to {@code true}.
+ * The comparison to {@link #maxAttempts(long)} will then be done with the number of subsequent attempts
+ * that failed without an {@link org.reactivestreams.Subscriber#onNext(Object) onNext} in between.
+ * <p>
+ * The {@link RetryBackoffSpec} is copy-on-write and as such can be stored as a "template" and further configured
+ * by different components without a risk of modifying the original configuration.
+ *
+ * @author Simon Baslé
+ */
+public final class RetryBackoffSpec implements Retry, Supplier<Retry> {
```

Comment:
It doesn't look like it needs to be `Supplier`, probably a left-over from earlier draft..